docs/includes/ksql-includes.rst

```diff
@@ -347,3 +435,443 @@ To enable JMX metrics, set ``JMX_PORT`` before starting the KSQL server:
+
+.. code:: bash
+
+    0 | City_35
+    1 | City_21
+    2 | City_47
+    3 | City_57
+    4 | City_17
+
+Press Ctrl-C to cancel the ``SELECT`` query. 
+
+
+.. __struct_support_02_end
+
+.. __stream_stream_join:
+
+.. __ss-join_01_start
+
+Stream-Stream join
+==================
+
+Using a stream-stream join, it is possible to join two *streams* of
+events on a common key. An example of this could be a stream of order
+events, and a stream of shipment events. By joining these on the order
+key, it is possible to identify any orders that have not yet shipped.
+
+First, populate the ``orders`` and ``shipments`` topics:
+
+.. __ss-join_01_end
+
+.. __ss-join_02_start
+
+Register both topics with KSQL:
+
+.. code:: sql
+
+    ksql> CREATE STREAM NEW_ORDERS (ORDER_ID INT, TOTAL_AMOUNT DOUBLE, CUSTOMER_NAME VARCHAR) \
+          WITH (KAFKA_TOPIC='new_orders', VALUE_FORMAT='JSON');
+
+    ksql> CREATE STREAM SHIPMENTS (ORDER_ID INT, SHIPMENT_ID INT, WAREHOUSE VARCHAR) \
+          WITH (KAFKA_TOPIC='shipments', VALUE_FORMAT='JSON');
+
+After each ``CREATE STREAM`` statement you should get the message: 
+
+.. code:: bash
+
+     Message
+    ----------------
+     Stream created
+    ----------------
+
+Query the data to confirm that it is present in the topics. 
+
+.. include:: ../includes/ksql-includes.rst
+    :start-after: _offsetreset_start
+    :end-before: _offsetreset_end
+
+For the ``NEW_ORDERS`` topic, run: 
+
+.. code:: sql
+
+    ksql> SELECT ORDER_ID, TOTAL_AMOUNT, CUSTOMER_NAME FROM NEW_ORDERS LIMIT 3;
+
+Your output should resemble:
+
+.. code:: bash
+
+    1 | 10.5 | Bob Smith
+    2 | 3.32 | Sarah Black
+    3 | 21.0 | Emma Turner
+
+For the ``SHIPMENTS`` topic, run: 
+
+.. code:: sql
+
+    ksql> SELECT ORDER_ID, SHIPMENT_ID, WAREHOUSE FROM SHIPMENTS LIMIT 2;
+
+Your output should resemble:
+
+.. code:: bash
+
+    1 | 42 | Nashville
+    3 | 43 | Palo Alto
+
+Run the following query, which will show orders with associated shipments, 
+based on a join window of 1 hours. This means that anything not yet shipped, 
+or shipped after the 1 hour period, will not return in this join:
+
+.. code:: sql
+
+    ksql> SELECT O.ORDER_ID, O.TOTAL_AMOUNT, O.CUSTOMER_NAME, \
+          S.SHIPMENT_ID, S.WAREHOUSE \
+          FROM NEW_ORDERS O \
+          LEFT JOIN SHIPMENTS S \
+            WITHIN 1 HOURS \
+            ON O.ORDER_ID = S.ORDER_ID;
+
+Your output should resemble:
+
+.. code:: bash
+
+    1 | 10.5 | Bob Smith | 42 | Nashville
+    2 | 3.32 | Sarah Black | null | null
+    3 | 21.0 | Emma Turner | 43 | Palo Alto
+
+Note that message with ``ORDER_ID=2`` has no corresponding
+``SHIPMENT_ID`` or ``WAREHOUSE`` - this is because there is no
+corresponding row on the shipments stream within the time window
+specified. 
+
+Press Ctrl-C to cancel the ``SELECT`` query and return to the KSQL prompt.
+
+We can now write a query to show us orders that have
+not yet shipped within this time window, by adding a ``WHERE`` clause:
```

Comment:
I'm not sure that we want to use this specific use case. Any time the ORDERS stream has a record for an order with an earlier timestamp than the corresponding record in the SHIPMENTS stream, the output stream will contain both a null shipment and non-null shipment for that order. So the filter query would return a false positive. This ordering of events would be the norm for something like shipments/orders where the shipment follows the order (maybe one day Alexa will be so advanced this won't always be the case 😄). I think for stream-stream join it would be simplest to use cases that rely on positive join results. For example, correlating ad impressions with orders to determine how effective an ad is. If we want to use negative join results, we should either pick a use case for which the right side is highly likely to be present when the left side arrives, or use something at the sink that takes into account "corrected" results.

Reply:
Yeah, I think this is what I hit in #1558 :-(

I'll try and re-work this example today.

Reply:
@rodesai updated - please review. thanks.